File: doc/author.mjs
#!/usr/bin/env node
// @gcu/cradle — doc capsule author (Node, dependency-free; stdlib zlib only).
// Turns a Markdown document (with optional YAML frontmatter) into a `!doc1+` capsule and a
// shareable cradle URL. The recipient opens the link; cradle renders it — no file, no host.
//
//   node author.mjs <file.md>                       # → prints the share URL
//   node author.mjs - --locale pt-BR                # read the doc from stdin
//   node author.mjs report.md --base https://gentropic.org/cradle/
//
// Mirror of author.py (same capsule scheme). Run `validate.mjs` first to catch problems.
// Capsule scheme: !doc1+<locale>\n<body>  →  raw-deflate  →  inline:deflate:<base64url>.
import { deflateRawSync } from "node:zlib";
import { readFileSync } from "node:fs";

const DEFAULT_BASE = "https://gentropic.org/cradle/";

export function makeDocCapsule(content, { locale = "en-US" } = {}) {
  const payload = `!doc1+${locale}\n${String(content)}`;
  const raw = deflateRawSync(Buffer.from(payload, "utf8"), { level: 9 });
  const b64url = raw.toString("base64").replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
  return "inline:deflate:" + b64url;
}
// fragment-escape per SPEC-capsule §6.4.1 (base64url has no % or space, so this is a no-op
// here, but keep it for correctness / parity with the q: scheme).
const fragmentEncode = (s) => s.replace(/%/g, "%25").replace(/ /g, "%20");

export function makeDocUrl(content, { locale = "en-US", base = DEFAULT_BASE } = {}) {
  return base.replace(/\/?$/, "/") + "#" + fragmentEncode(makeDocCapsule(content, { locale }));
}

function cli(argv) {
  const a = argv.slice(2);
  let file = null, locale = "en-US", base = DEFAULT_BASE;
  for (let i = 0; i < a.length; i++) {
    if (a[i] === "--locale") locale = a[++i];
    else if (a[i] === "--base") base = a[++i];
    else if (!a[i].startsWith("--")) file = a[i];
  }
  if (!file) { process.stderr.write("usage: author.mjs <file.md|-> [--locale en-US] [--base URL]\n"); process.exit(2); }
  const content = file === "-" ? readFileSync(0, "utf8") : readFileSync(file, "utf8");
  const cap = makeDocCapsule(content, { locale });
  const url = base.replace(/\/?$/, "/") + "#" + fragmentEncode(cap);
  process.stdout.write(url + "\n");
  const ub = url.length;
  const nfc = [["NTAG213", 144], ["NTAG215", 504], ["NTAG216", 888]].map(([n, c]) => `${n} ${ub <= c ? "✓" : "✗"}`).join(" · ");
  process.stderr.write(`capsule ${cap.length} B · URL ${ub} B · NFC: ${nfc}\n`);
}

if (process.argv[1] && process.argv[1].endsWith("author.mjs")) cli(process.argv);

File: doc/author.py
#!/usr/bin/env python3
"""@gcu/cradle — doc capsule author (Python, stdlib only: zlib + base64).

Turns a Markdown document (with optional YAML frontmatter) into a `!doc1+` capsule and a
shareable cradle URL. The recipient opens the link; cradle renders it — no file, no host.

    python author.py report.md                       # -> prints the share URL
    cat report.md | python author.py - --locale pt-BR
    python author.py report.md --base https://gentropic.org/cradle/

Mirror of author.mjs (same capsule scheme). Run validate.py first to catch problems.
Capsule scheme: !doc1+<locale>\\n<body>  ->  raw-deflate  ->  inline:deflate:<base64url>.
"""
import sys
import zlib
import base64

DEFAULT_BASE = "https://gentropic.org/cradle/"


def make_doc_capsule(content, locale="en-US"):
    payload = ("!doc1+%s\n%s" % (locale, content)).encode("utf-8")
    co = zlib.compressobj(9, zlib.DEFLATED, -zlib.MAX_WBITS)   # raw deflate (no zlib header)
    raw = co.compress(payload) + co.flush()
    b64url = base64.urlsafe_b64encode(raw).decode("ascii").rstrip("=")
    return "inline:deflate:" + b64url


def fragment_encode(s):   # SPEC-capsule §6.4.1 (no-op for base64url; kept for parity)
    return s.replace("%", "%25").replace(" ", "%20")


def make_doc_url(content, locale="en-US", base=DEFAULT_BASE):
    cap = make_doc_capsule(content, locale)
    return base.rstrip("/") + "/#" + fragment_encode(cap)


def _cli(argv):
    file = None
    locale = "en-US"
    base = DEFAULT_BASE
    i = 0
    rest = argv[1:]
    while i < len(rest):
        if rest[i] == "--locale":
            i += 1; locale = rest[i]
        elif rest[i] == "--base":
            i += 1; base = rest[i]
        elif not rest[i].startswith("--"):
            file = rest[i]
        i += 1
    if not file:
        sys.stderr.write("usage: author.py <file.md|-> [--locale en-US] [--base URL]\n")
        sys.exit(2)
    content = sys.stdin.read() if file == "-" else open(file, "r", encoding="utf-8").read()
    cap = make_doc_capsule(content, locale)
    url = base.rstrip("/") + "/#" + fragment_encode(cap)
    sys.stdout.write(url + "\n")
    ub = len(url)
    nfc = " · ".join("%s %s" % (n, "OK" if ub <= c else "X") for n, c in (("NTAG213", 144), ("NTAG215", 504), ("NTAG216", 888)))
    sys.stderr.write("capsule %d B · URL %d B · NFC: %s\n" % (len(cap), ub, nfc))


if __name__ == "__main__":
    _cli(sys.argv)
